### PR TITLE
Sync: Full sync the term relationships

### PR DESCRIPTION
**Pull request by tyxla**
This implements syncing of the term relationships in full sync, which currently we don't seem to be doing.

Turns out it is a bit specific use case, because that table the only one that doesn't have a unique ID column. So to be able to still chunk properly like we do for other sync modules, we use 2 columns, and wrap the functionality in a new sync module, `Term_Relationships`.

We're introducing some tests to verify that missing or obsolete term relationships are synced, and add some tests to the full sync to ensure that it's properly syncing all term relationships, and is properly chunking them when there are a lot of them to sync.

#### Changes proposed in this Pull Request:
* Sync: Full sync the term relationships

#### Is this a new feature or does it add/remove features to an existing part of Jetpack?
* Part of the effort to fix term syncing and the overall state of sync.

#### Testing instructions:
* Apply D30623-code to your WP.com sandbox.
* Apply this PR to your local Jetpack dev env.
* Enable `WP_DEBUG_LOG` on your site if it's disabled.
* Silently create a term relationship between a post and a term on your site (see below for how to do that).
* Trigger a full sync.
* Make sure the term relationship got synced to the cache site.
* Verify full and incremental sync for other items still works.
* Try testing with some other use cases you can think of?
* Verify the tests pass.
* Verify there are no errors.

#### Additional testing help
* Silent creation term relationship can be done with this piece of code:
```
global $wpdb;
$wpdb->insert(
	$wpdb->term_relationships,
	array(
		'object_id'        => 1234,
		'term_taxonomy_id' => 123,
		'term_order'       => 0,
	),
	array(
		'%d',
		'%d',
		'%d',
	)
);
```

#### Proposed changelog entry for your changes:
* Sync: Full sync the term relationships

### CODE REVIEW

**Comment by tyxla** [comment]
This is ready for review now.

File: packages/sync/src/modules/Term_Relationships.php
Comment (tyxla) [inline packages/sync/src/modules/Term_Relationships.php]: The bit in the `while` is kind of the same as the one in `Automattic\Jetpack\Sync\Modules\Module::enqueue_all_ids_as_action`, so we can eventually refactor them to share the same method, but it feels like this can be done in a follow-up PR, hence the `@todo` in the docblock.

**Review by roccotripaldi — APPROVED** [review]
This is looking good. Thanks for working on it.

One thing is missing from the testing instructions:
- you need to set JETPACK__API_BASE to your sandbox in jetpack.php

After testing, i realized we had to make a small change to D30623.

It works well when you silently add a term_relationship: I see it appear in the cache site after a full sync.

When i silently delete a term_relationship, the entry remains on the cache site after a full sync.

I think we can fix that on the server-side, so you have my approval.

**Comment by roccotripaldi** [comment]
This PR is not needed any more. In D30684-code, we ensured that we were properly getting term relationships during a full sync.

**Comment by lezama** [comment]
Not 100% sure about closing it, I think it might be useful to have it anyways, even if full syncing posts should fix terms, maybe "fullsyncing" just this table could be faster in some cases, plus we might need some of the logic here for the validator, what do you think?

**Comment by gravityrail** [comment]
@lezama @tyxla @roccotripaldi 

Let's merge this, even if we don't use it right now, _just_ in case we discover we need to revert to this approach during testing with very large sites on VIP.

**Review by zinigor — APPROVED** [review]
Although this might need a lint pass after, let's do it in another PR.

**Pull request merged**
